contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -727,18 +832,131 @@ cannot find a node for the pod, then it will encounter scheduling failure and
+8. **NEW:** Still in the assume func, it also check if any provisioning operations
+   is needed. If so, it will update the PVC cache (store provisioned PVCs
+   per StorageClass), to mark the capacity occupied.
+9. If PVC binding or provisioning is required, we do NOT AssumePod. Instead,
+   a new bind function, BindPVCs, will be called asynchronously,
+   passing in the selected node. The bind function will prebind the PV to the PVC,
+   or trigger dynamic provisioning (NEW). Then, it always sends the Pod through the
+   scheduler again.
+10. When a Pod makes a successful scheduler pass once all PVCs are bound/provisioned,
+   the scheduler assumes and binds the Pod to a Node.
+11. Kubelet starts the Pod.
+
+#### Scheduler Changes
+
+##### Predicate
+
+In the predicate function of VolumeBindingChecker, if the scheduler fail
+to find matching volumes for an unbound claim, instead of return false as it does now,
+the scheduler tries to trigger provisioning:
+1. If the PVC has an anotation of `annStorageProvisioner`, it means provisioning
+   has already been triggered on the PVC, skip following steps for the PVC,
+   to wait for the provisioning to finish
+2. Fetch the StorageClass according to the PVC.
+3. Check if provisioner is specified in StorageClass, if not, return false directly
+4. Check if the node could satisfy the topology requirement in StorageClass, if not,
+   return false directly
+5. Check if `CapacityTopologyKey` is specified in the StorageClass, if not,
+   we assume the storage is of infinite capacity, and skip the check on capacity
+6. If specified, go through the labels on the node, to see if any label key could
+   match the CapacityTopologyKey, if not, we assume the node not match the capacity
+   requirement.
+7. If match, try to find capacity value with node label value as the capacity
+   record key (e.g. node1 has label "kubernetes.io/hostname=node1", storageclass S1
+   has CapacityTopologyKey "capacityTopologyKey=kubernetes.io/hostname", then the
+   scheduler tries to find capacity information with "node1" as key of capacity map
+   in StorageClass, S1.status.capacity["node1"]). If capacity not found, or the
+   allocatable storage (recorded capacity reduces that already used) is smaller than
+   that required in PVC, we assume the node not match the capacity requirement.
+8. If capacity requirement satisfied, Temporarily cache provisioned PVCs per Node,
+   for fast processing later in the priority and bind function
+9. Continue to process next claim of the pod
+
+##### Assume
+
+For better scheduler performance, we'll assume that the provisioning
+will likely succeed, and update the cache with the PVC provisioned,
+assuming the part of the capacity occupied, so that scheduler could
+continue processing privsioning before actual API update on PVC
+(update in VolumeName field that indicates provisioning finished).
+
+In addition to `bindingRequired` that function `AssumePodVolumes`currently
```

Comment:
I'm not sure if an additional provisioning required argument is needed.  Can bindingRequired be used for both?

Reply:
Yeah, you are right, I meant to change the part in a simple way, but was not aware of the fact that it's related to change in interfaces, which should be very cautious.
We could manage it with existing argument, just some additional logic in current check on `bindingRequired`